bootstrap/util/configowner.go

```diff
@@ -77,20 +79,36 @@ func (co ConfigOwner) IsControlPlaneMachine() bool {
 // GetConfigOwner returns the Unstructured object owning the current resource.
 func GetConfigOwner(ctx context.Context, c client.Client, obj metav1.Object) (*ConfigOwner, error) {
 	for _, ref := range obj.GetOwnerReferences() {
-		refgvk, err := schema.ParseGroupVersion(ref.APIVersion)
+		refGV, err := schema.ParseGroupVersion(ref.APIVersion)
 		if err != nil {
 			return nil, errors.Wrapf(err, "failed to parse GroupVersion from %q", ref.APIVersion)
 		}
+		refGVK := refGV.WithKind(ref.Kind)
 
-		if (ref.Kind == "Machine" || ref.Kind == "MachinePool") &&
-			refgvk.Group == clusterv1.GroupVersion.Group {
-			return GetOwnerByRef(ctx, c, &corev1.ObjectReference{
-				APIVersion: ref.APIVersion,
-				Kind:       ref.Kind,
-				Name:       ref.Name,
-				Namespace:  obj.GetNamespace(),
+		allowedGKs := []schema.GroupKind{
+			{
+				Group: clusterv1.GroupVersion.Group,
+				Kind:  "Machine",
+			},
+		}
+
+		if feature.Gates.Enabled(feature.MachinePool) {
```

Comment:
Since this is an externally available API, how do we expect an out-of-tree bootstrap provider to consume this? Do we need to add it to the v1alpha2 to v1alpha3 migration docs that using this library will also require also embedding the core feature gate code to support experimental features?

Reply:
this is a problem even for kubeadm bootstrapper, given that we're shipping different binaries

Reply:
🤔 might need more time to think about this